content/en/docs/concepts/services-networking/endpoint-slices.md

```diff
@@ -93,6 +93,44 @@ EndpointSlices support three address types:
 * IPv6
 * FQDN (Fully Qualified Domain Name)
 
+### Conditions
+
+The EndpointSlice API stores conditions about endpoints that may be useful for consumers.
+The three conditions are `ready`, `serving`, and `terminating`.
+
+#### Ready
+
+`Ready` is a condition that maps to a pod's `Ready` condition. A running pod with the `Ready` condition set to `True`
+should have this EndpointSlice condition also set to `true`. For compatibility reasons, `ready` is NEVER `true` when a
+pod is terminating. Consumers should refer to the `serving` condition to inspect pod readiness for terminating pods.
+The only exception to this rule is for Services with `spec.publishNotReadyAddresses` set to `true`. Endpoints for these
+Services will always have `ready` condition.
+
+#### Serving
+
+{{< feature-state for_k8s_version="v1.20" state="alpha" >}}
+
+`Serving` is identical to the `Ready` condition, except it does not account for terminating states. Consumers of the
```

Comment:
Does this include the adjustment for `spec.publishNotReadyAddresses`?

Reply:
Interestingly, I don't think `serving` is always true for `spec.publishNotReadyAddresses`. Do we need to change that for beta?

I see value in keeping `serving` distinct from `publishNotReadyAddresses` but I need to catch up on all the context behind this field.

Reply:
This feels like a really tricky one. The [comment for `publishNotReadyAddresses`](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go#L3689-L3691) says:

```
The Kubernetes controllers that generate Endpoints and EndpointSlice resources for Services interpret this to mean that all endpoints are considered "ready" even if the Pods themselves are not.
```

As it reads right now, I think it would make sense for the behavior of `serving` to mirror that of `ready` here. It looks like @JeremyOT made the last update to this in https://github.com/kubernetes/kubernetes/pull/90783, any thoughts here? Also interested in perspective from @thockin on how we should treat `serving` here.

Reply:
If that is the behavior we want, is there any way we could still get this into 1.20 as a bug fix?

Reply:
I think it would be great to know if pods were in fact ready/serving regardless of pNRA.

I don't think the current pNRA behavior is intuitive which is why I updated the docs to explain what actually happens. publishNotReadyAddresses should really be named makeEndpointsAlwaysReady (there's no not-ready endpoints to publish, they're all ready).

It seems like a win to have "serving" always provide an accurate view of the endpoint - especially if that's what it's already doing.

Reply:
That makes sense, it is a super confusing field. Maybe we'll need another update to the docs around that field once `Serving` becomes available.

Reply:
Thanks folks, fwiw I feel that `serving` actually removes the need for `spec.publishNotReadyAddreses` altogether. I think that field was mostly added due to old behavior with Endpoints only including ready endpoints and excluding anything else.

Reply:
Isn't pNRA focused on DNS? Endpoints include everything, but DNS still only publishes ready endpoints by default. In an ideal world where we didn't have legacy behavior to account for I think pNRA would just be a signal to the DNS implementation and not affect endpoint readiness at all.

Reply:
Good point